wcfsetup/install/files/lib/system/label/LabelPicker.class.php

```diff
@@ -0,0 +1,141 @@
+     * for items that are not assigned any label of a label group.
+     */
+    public readonly bool $invertible;
+
+    /**
+     * Name of the hidden input field.
+     */
+    public string $name = 'labelIDs';
+
+    public readonly ViewableLabelGroup $labelGroup;
+
+    private string $elementID;
+
+    private int $selected = 0;
+
+    public function __construct(ViewableLabelGroup $labelGroup, bool $invertible)
+    {
+        $this->labelGroup = $labelGroup;
+        $this->invertible = $invertible;
+    }
+
+    /**
+     * Sets the selected label of this label picker by providing its id. The
+     * value `0` indicates that no selection is to be made and  `-1` inverts the
+     * selection of the labels.
+     */
+    public function setSelectedValue(int $selected): void
+    {
+        if ($selected === 0) {
+            $this->selected = 0;
+        } else if ($selected === -1) {
+            $this->selected = $this->invertible ? -1 : 0;
+        } else {
+            $this->selected = $this->labelGroup->isValid($selected) ? $selected : 0;
+        }
+    }
+
+    /**
+     * The returned value can be `0` to indicate that no selection has been
+     * made, `-1` to indicate that the selection should be inverted or the id
+     * of the selected label.
+     */
+    public function getSelectedValue(): int
+    {
+        return $this->selected;
+    }
+
+    /**
+     * Returns true if a label has been selected or if the selection has been
+     * inverted.
+     */
+    public function hasSelection(): bool
+    {
+        return $this->selected !== 0;
+    }
+
+    /**
+     * Generates the HTML element for the label picker.
+     */
+    public function toHtml(): string
+    {
+        $labels = [];
+        foreach ($this->labelGroup as $label) {
+            \assert($label instanceof Label);
+
+            $labels[] = [$label->labelID, $label->render()];
+        }
+
+        return \sprintf(
+            <<<'EOT'
+                <woltlab-core-label-picker
+                    selected="%d"
+                    id="%s"
+                    title="%s"
+                    labels="%s"
+                    name="%s"
+                    data-group-id="%d"
+                    %s
+                ></woltlab-core-label-picker>
+            EOT,
+            $this->selected,
+            $this->getElementID(),
+            StringUtil::encodeHTML($this->labelGroup->getTitle()),
+            StringUtil::encodeHTML(JSON::encode($labels)),
+            StringUtil::encodeHTML($this->name),
+            $this->labelGroup->groupID,
+            $this->invertible ? 'invertible' : '',
+        );
+    }
+
+    /**
+     * Returns the unique element id of this label picker.
+     */
+    public function getElementID(): string
+    {
+        if (!isset($this->elementID)) {
+            $this->elementID = \sprintf(
+                '%s_labelGroup%d',
+                \substr(\md5($this->name), 0, 8),
```

Comment:
Is it important that the elementID is reproducible or does it just need to be unique?

If it does not need to be reproducible, just use: `bin2hex(random_bytes(16))`.

If it does need to be reproducible, the `md5()` call is an unnecessary red flag and can still conflict, e.g. when more than one form exists on a page?